rclpy/src/rclpy/_rclpy.c

```diff
@@ -155,14 +155,58 @@ rclpy_trigger_guard_condition(PyObject * Py_UNUSED(self), PyObject * args)
  * Raises RuntimeError if rcl could not be initialized
  */
 static PyObject *
-rclpy_init(PyObject * Py_UNUSED(self), PyObject * Py_UNUSED(args))
+rclpy_init(PyObject * Py_UNUSED(self), PyObject * args)
 {
-  // TODO(esteve): parse args
-  rcl_ret_t ret = rcl_init(0, NULL, rcl_get_default_allocator());
-  if (ret != RCL_RET_OK) {
-    PyErr_Format(PyExc_RuntimeError,
-      "Failed to init: %s", rcl_get_error_string_safe());
-    rcl_reset_error();
+  // Expect one argument which is a list of strings
+  PyObject * pyargs;
+  if (!PyArg_ParseTuple(args, "O", &pyargs)) {
+    // Exception raised
+    return NULL;
+  }
+
+  pyargs = PySequence_List(pyargs);
+  if (NULL == pyargs) {
+    // Exception raised
+    return NULL;
+  }
+  Py_ssize_t pysize_num_args = PyList_Size(pyargs);
+  if (pysize_num_args > INT_MAX) {
+    PyErr_Format(PyExc_OverflowError, "Too many arguments");
+    return NULL;
+  }
+  int num_args = (int)pysize_num_args;
+
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  char ** arg_values = allocator.allocate(sizeof(char *) * num_args, allocator.state);
+  if (NULL == arg_values) {
+    PyErr_Format(PyExc_MemoryError, "Failed to allocate space for arguments");
+    Py_DECREF(pyargs);
+    return NULL;
+  }
+
+  bool have_args = true;
+  for (int i = 0; i < num_args; ++i) {
+    // Returns borrowed reference, do not decref
+    PyObject * pyarg = PyList_GetItem(pyargs, i);
+    if (NULL == pyarg) {
+      have_args = false;
+      break;
+    }
+    // Borrows a pointer, do not free arg_values[i]
+    arg_values[i] = PyUnicode_AsUTF8(pyarg);
+  }
```

Comment:
can we just skip this entire block (L180 to 197) when `num_args` is 0 and pass a `NULL` pointer to `rcl_init` instead?

Reply:
Block skipped when no args in 93c634bf20fb57def7ca3ff7517553b53d99caa3